pep-0613.rst

```diff
@@ -0,0 +1,223 @@
+determined: if a top level assignment is unannotated, and the assigned value is
+a valid type, then the name being assigned to is a valid type alias. Otherwise,
+that name is simply a global value that cannot be used as a type hint.
+
+These implicit type alias declaration rules create confusion when type aliases
+involve forward references, invalid types, or violate other restrictions
+enforced on type alias declaration. Because the distinction between an
+unannotated value and a type alias is implicit, ambiguous or incorrect type
+alias declarations implicitly default to a valid value assignment. This creates
+expressions that are impossible to express as type aliases and punts error
+diagnosis of malformed type aliases downstream.
+
+The following examples each include an illustration of some of the suboptimal
+or confusing behaviors resulting from existing implicit alias declarations.
+We also introduce explicit aliases of the format :code:`TypeName: TypeAlias = Expression`
+here for the sake of comparison, but the syntax is discussed in further detail
+in later sections.
+
+Forward References:
+*******************
+
+::
+
+  MyType = "ClassName"
+  def foo() -> MyType: ...
+
+This code snippet should not error so long as :code:`ClassName` is defined
+later on. However, a type checker is forced to throw spurious errors that
+(1) :code:`MyType` is an unannotated global string, and (2) :code:`MyType`
+is not a valid type.
+
+::
+
+  MyType: TypeAlias = “ClassName”
+  def foo() -> MyType: ...
+
+Explicit aliases remove ambiguity so neither of the above errors will be
+thrown. Additionally, if something is wrong with :code:`ClassName`
+(i.e., it’s not actually defined later), the type checker can throw an error.
+
+
+Error Messaging:
+****************
+
+::
+
+  MyType1 = InvalidType
+  MyType2 = MyGeneric(int)  # i.e., intention was MyGeneric[int]
+
+A type checker should warn on this code snippet that :code:`InvalidType` is not
+a valid type, and therefore cannot be used to annotate an expression or to
+construct a type alias. Instead, type checkers are forced to throw spurious
+errors that (1) :code:`MyType` is a global expression missing an annotation,
+and (2) :code:`MyType` is not a valid type in all usages of :code:`MyType`
+across the codebase.
+
+::
+
+  MyType1: TypeAlias = InvalidType
+  MyType2: TypeAlias = MyGeneric(int)
+
+With explicit aliases, the type checker has enough information to error on the
+actual definition of the bad type alias, and explain why: that :code:`MyGeneric(int)`
+and `InvalidType` are not valid types. When the value expression is no longer
+evaluated as a global value, unactionable type errors on all usages of :code:`MyType`
+across the codebase can be suppressed.
+
+Scope Restrictions:
+*******************
+
+::
+
+  x = ClassName
+  def foo() -> None:
+    x = ClassName
+
+The outer :code:`x` is a valid type alias, but type checkers must error if the
```

Comment:
It wasn't obvious to me why type checkers must error here until I got to the explanation at the very end of the section, so I'd suggest moving that explanation here: "type checkers must error if the inner `x` is ever used as a type <b>because type aliases cannot be defined inside a nested scope</b>."

Reply:
Hi! I'm curious about the rationale behind this limitation.
Here's a sample snippit and how it works currently.

```python
from typing import NewType
from typing_extensions import TypeAlias

class C:
    X = NewType('X', int)
    Y1 = X  # mypy - error: Variable "test.C.Y1" is not valid as a type  pyright - seems to work
    Y2: TypeAlias = X  # mypy - seems to work  pyright - error: A TypeAlias can be defined only within a module scope

x: C.X
y1: C.Y1
y2: C.Y2
```

I have a use case like this where there is a type `C.Y` that I'm trying to rename to `C.X` in a backward compatible way. Type alias would be perfect, but PEP613 disallows it (and behavior between existing typecheckers vary). Other recommendations on how to do this welcome.

The PEP mentions how non-explicit alias declarations are confusing in nested scope. From first principles, something like `Y1 = X` seems ok to allow as a type alias. I may be overlooking something.

Reply:
From https://mail.python.org/archives/list/typing-sig@python.org/thread/MWRJOBEEEMFVXE7CAKO7B4P46IPM4AN3/ (@ilevkivskyi)

> ```
> class C: ...
> 
> x = C  # this is a type alias from mypy point of view
> class D:
>     x = C  # this is a variable with type Type[C]
> ```
> This default behaviour is kind of a heuristics that matches most typical
> use in each case, but I think it would be nice to have a simple way to be
> explicit either way:


If provided explicitly (like `Y2: TypeAlias`) - it seems reasonable to be allowable. The use case of backwards-compatible rename of an inner type is compelling to me.

If this makes sense, I'm happy to put some legwork in to update/clarify the PEP

Reply:
Probably better to bring this up on typing-sig. I don't have a strong opinion on this either way, your suggestion seems reasonable.